[Coupons] list improvements

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #6468 Closes: #6470 Closes #6338

### Description
This PR adds the following functionalities to the coupons list:
- Swipe-to-refresh.
- A loading indicator when loading the next page.
- An error Snackbar if the loading of the next page fails.
- Fixes a bug that resulted in #6338 

It adds also some unit tests for the CouponsListViewModel.

### Testing instructions
For all tests, make sure the coupons beta feature flag is enabled
#### Pagination loading indicator
1. Open coupons list.
2. Scroll down to trigger loading of next page.
3. Confirm a circular progress indicator is shown at the end of the list during loading.
4. Confirm the loading indicator is hidden when loading is completed.

#### Swipe to refresh
1. Open coupons list.
2. Swipe down to trigger refresh.
3. Confirm the list is refreshed, and the indicator is hidden when it's completed.

#### Error snackbar
1. Enable airplane mode.
2. Make some actions: swipe to refresh or search.
3. Confirm a snackbar is shown with an error message.

### Images/gif

https://user-images.githubusercontent.com/1657201/169901979-af3f3f99-533f-49e1-83b0-62f30571181c.mov

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Introduce a loading state enum

hichamboushaba committed:
Use OptIn instead of FlowPreview to remove warnings

hichamboushaba committed:
Use explicit imports

hichamboushaba committed:
Add a loading indicator for loading next pages

hichamboushaba committed:
Introduce a delay for loading state changes

This is needed since we are observing the DB changes, so we need to make sure list is fetched before setting state to Idle.

hichamboushaba committed:
Add some padding around the loading indicator

hichamboushaba committed:
Fix search view behavior

hichamboushaba committed:
Add swipe-to-refresh

hichamboushaba committed:
Cleanup baseline

hichamboushaba committed:
Show error when fetching/searching fails

hichamboushaba committed:
Add unit tests for the viewmodel

hichamboushaba committed:
Remove unneeded stubbing

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/CouponListFragment.kt:
```diff
@@ -107,32 +117,34 @@ class CouponListFragment : BaseFragment(R.layout.fragment_coupon_list) {
         }
-        searchMenuItem.setOnActionExpandListener(object : OnActionExpandListener {
-            override fun onMenuItemActionExpand(item: MenuItem?): Boolean {
+        val textQueryListener = object : OnQueryTextListener, SearchView.OnQueryTextListener {
+            override fun onQueryTextSubmit(query: String?): Boolean {
                 if (isAdded) {
-                    viewModel.onSearchStateChanged(open = true)
+                    viewModel.onSearchQueryChanged(query.orEmpty())
                 }
                 return true
             }
 
-            override fun onMenuItemActionCollapse(item: MenuItem?): Boolean {
+            override fun onQueryTextChange(newText: String?): Boolean {
                 if (isAdded) {
-                    viewModel.onSearchStateChanged(open = false)
+                    viewModel.onSearchQueryChanged(newText.orEmpty())
                 }
                 return true
             }
-        })
-        searchView.setOnQueryTextListener(object : OnQueryTextListener, SearchView.OnQueryTextListener {
-            override fun onQueryTextSubmit(query: String?): Boolean {
+        }
+        searchMenuItem.setOnActionExpandListener(object : OnActionExpandListener {
+            override fun onMenuItemActionExpand(item: MenuItem?): Boolean {
                 if (isAdded) {
-                    viewModel.onSearchQueryChanged(query.orEmpty())
+                    viewModel.onSearchStateChanged(open = true)
+                    searchView.setOnQueryTextListener(textQueryListener)
                 }
                 return true
             }
 
-            override fun onQueryTextChange(newText: String?): Boolean {
+            override fun onMenuItemActionCollapse(item: MenuItem?): Boolean {
                 if (isAdded) {
-                    viewModel.onSearchQueryChanged(newText.orEmpty())
+                    searchView.setOnQueryTextListener(null)
+                    viewModel.onSearchStateChanged(open = false)
```
These changes fixe #6338

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/CouponListViewModel.kt:
```diff
@@ -29,36 +38,38 @@ class CouponListViewModel @Inject constructor(
-    ) { coupons, isLoading, searchQuery ->
+        flow2 = loadingState.withIndex()
+            .debounce {
+                if (it.index != 0 && it.value == LoadingState.Idle) {
+                    // When resetting to Idle, wait a bit to make sure the coupons list has been fetched from DB
+                    LOADING_STATE_DELAY
```
Since we are observing the DB for reflecting the data changes, there is a slight delay between marking the loading as finished, and getting the data from the DB, this delay just makes sure the DB changes get to the UI before marking the loading as finished, to prevent flickering on the list.

hichamboushaba committed:
Show an error message when loading next page fails

hichamboushaba committed:
Add unit tests for error scenarios

hichamboushaba committed:
Fix the query check for defining the error message

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/CouponListViewModel.kt:
```diff
@@ -94,26 +109,47 @@ class CouponListViewModel @Inject constructor(
-                        couponListHandler.fetchCoupons(searchQuery = it)
+                        couponListHandler.fetchCoupons(
+                            searchQuery = query,
+                            forceRefresh = index == 0 && query == null // Force refresh on initial loading
```
This is working well, I was just thinking it's a bit confusing that the initial non-search fetch is executed inside the `monitorSearchQuery()` method. I took me a moment to find it after seeing it was removed from `init()`. I like how it's automatically integrated, though. Maybe mentioning initial load in the method name could improve this?

hichamboushaba replied:
I actually felt the same, and nice job challenging me on this.
I did a small refactoring to this to bring back individual block of initial fetch, check 6ebc0b16644c98dba19f38aed8905b8e6f7b27e8 and let me know what you think.

0nko replied:
Thanks for the updates, Hicham!

0nko reviewed (approved):
Looking great. 👍  I left a very minor comment in the code, but feel free to ignore it.

hichamboushaba committed:
Refactor logic of initial fetching of coupons

This decouples the initial fetching from the search status, to make the purpose clearer

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/CouponListViewModel.kt:
```diff
@@ -118,29 +121,41 @@ class CouponListViewModel @Inject constructor(
+                .withIndex()
+                .filterNot {
+                    // Skip initial value to avoid double fetching coupons
+                    it.index == 0 && it.value == null
```
I'm checking the `value` and the `index` both, as we save the search query to the saved state, and we don't want to skip the initial value when restoring from it.
